File: src/backend/upload/annotator.py
from .serializables import Corpus, Paragraph, Token

import itertools
import string
import re
from typing import List, Tuple

TOKEN_UNKNOWN = "!UNKNOWN"

def get_annotator(annotator_name):
	if annotator_name and annotator_name.startswith("chatgpt_"):
		from .chatgpt_annotator import ChatgptAnnotator
		return ChatgptAnnotator(annotator_name=annotator_name)
	else:
		return Annotator()

class Annotator:
	def __init__(self):
		self.annotator_name = "DUMMYGLOSS"

	def annotate(self, p: Paragraph, lang_from: str, lang_to: str, **kwargs):
		self.lang_from = lang_from
		self.lang_to = lang_to

		if not p.is_delimiter:
			if p.tokens != []:
				self.put_gloss(p)
			p.pstate = "ANNOTATED"

			p.annotator_info_obj.lang_from = self.lang_from
			p.annotator_info_obj.lang_to = self.lang_to
			p.annotator_info_obj.annotator_name = self.annotator_name

	def reannotate(self, p: Paragraph, lang_from: str, lang_to: str, target_tokens: List[str], **kwargs):
		self.lang_from = lang_from
		self.lang_to = lang_to

		if p.is_delimiter:
			return
		
		self.reput_gloss(p, target_tokens)
		
	def put_gloss(self, p: Paragraph):
		for token in p.tokens:
			if token.is_delimiter:
				continue
			token.gloss = f"{self.annotator_name}({token.txt})"

		p.annotator_info = f"dummy_`{self.lang_from}`_`{self.lang_to}`" #obsolete
		
	def reput_gloss(self, p: Paragraph, target_tokens: List[int]):
		if target_tokens is not None and target_tokens != []:
			assert type(target_tokens[0]) == int #a test code

		for i, t in enumerate(p.tokens):
			if i not in target_tokens:
				continue

			t.gloss = f"{self.annotator_name}_reannotated_`{self.lang_from}`_`{self.lang_to}`"

File: src/backend/gs_index/src/annotator.js
//Follows annotator.py

export { Annotator, TOKEN_UNKNOWN };

const TOKEN_UNKNOWN = "!UNKNOWN";

class Annotator {
	constructor() {
		this.annotator_name = "DUMMYGLOSS";
	}

	async annotate(p, lang_from, lang_to) {
		this.lang_from = lang_from;
		this.lang_to = lang_to;

		if (!p.is_delimiter) {
			if (p.tokens && p.tokens.length > 0)
				await this.put_gloss(p);
			p.state = "ANNOTATED";

			p.annotator_info_obj.lang_from = this.lang_from
			p.annotator_info_obj.lang_to = this.lang_to
			p.annotator_info_obj.annotator_name = this.annotator_name
		}
	}

	async reannotate(p, lang_from, lang_to, target_tokens) {
		this.lang_from = lang_from;
		this.lang_to = lang_to;

		if (p.is_delimiter)
			return;

		await this.reput_gloss(p, target_tokens);
	}

	async put_gloss(p) {
		for (const token of p.tokens) {
			if (token.is_delimiter)
				continue;
			token.gloss = `${this.annotator_name}(${token.txt})`
		}
	}

	async reput_gloss(p, target_tokens) {
		for (const [i, t] of p.tokens.entries()) {
			if (!target_tokens.includes(i))
				continue;

			t.gloss = `${this.annotator_name}_reannotated_\`${this.lang_from}\`_\`${this.lang_to}\``
		}
	}
}